raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java

```diff
@@ -1436,6 +1441,47 @@ public void testInvalidFetchRequest() throws Exception {
         context.assertSentFetchPartitionResponse(Errors.INVALID_REQUEST, epoch, OptionalInt.of(localId));
     }
 
+    // This test mainly focuses on whether the leader state is correctly updated under different fetch version.
+    @ParameterizedTest
+    @ApiKeyVersionsSource(apiKey = ApiKeys.FETCH)
+    public void testLeaderStateUpdateWithDifferentFetchRequestVersions(short version) throws Exception {
+        int localId = 0;
+        int otherNodeId = 1;
+        int epoch = 5;
+        Set<Integer> voters = Utils.mkSet(localId, otherNodeId);
+
+        RaftClientTestContext context = RaftClientTestContext.initializeAsLeader(localId, voters, epoch);
+
+        // First, test with a correct fetch request.
+        FetchRequestData fetchRequestData = context.fetchRequest(epoch, otherNodeId, 1L, epoch, 0);
+        FetchRequestData downgradedRequest = new FetchRequest.SimpleBuilder(fetchRequestData).build(version).data();
+        context.deliverRequest(downgradedRequest);
+        context.client.poll();
+        context.assertSentFetchPartitionResponse(Errors.NONE, epoch, OptionalInt.of(localId));
+        assertEquals(1L, context.log.highWatermark().offset);
+
+        // Next, create a fetch request that is designed to be divergent.
```

Comment:
I am not entirely convinced by the second part of the test. Basically, it relies on the fact that a fetch request with a diverging offset goes into the purgatory and is completed when the offset if finally available. I actually wonder if this is a bug because I would have expected the fetch request to be competed immediately in this case.

I can't really think of a good way to test the second part so I would just remove it.